src/NHibernate.Test/NHSpecificTest/NH3596/CachingWithTransformerTests.cs

```diff
@@ -0,0 +1,175 @@
+			{
+				var parentRoles = new List<Role>()
+                {
+                    new Role() { Name = "Admin", Children = null, Parent = null },
+                    new Role() { Name = "Manager", Children = null, Parent = null },
+                    new Role() { Name = "Support", Children = null, Parent = null }
+                };
+
+				var childRoles = new List<Role>()
+                {
+                    new Role() { Name = "Manager-Secretary", Children = null, Parent = parentRoles.FirstOrDefault(x => x.Name == "Manager") },
+                    new Role() { Name = "Superviser", Children = null, Parent = parentRoles.FirstOrDefault(x => x.Name == "Manager") }
+                };
+
+				foreach (var parentRole in parentRoles)
+				{
+					parentRole.Children = new List<Role>(childRoles.Where(x => x.Parent == parentRole));
+
+					session.Save(parentRole);
+				}
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.Delete("from System.Object");
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		protected override void Configure(Cfg.Configuration configuration)
+		{
+			base.Configure(configuration);
+
+			configuration.Cache(x =>
+								{
+									x.Provider<HashtableCacheProvider>();
+									x.UseQueryCache = true;
+								});
+		}
+
+		[Test]
+		public void UsingQueryOverToFutureWithCacheAndTransformerDoesntThrow()
+		{
+			//NH-3596
+			using (var session = this.OpenSession())
+			using (session.BeginTransaction())
+			{
+				Role children = null;
+				Role parent = null;
+
+				//store values in cache
+				session
+					.QueryOver<Role>()
+					.Where(x => x.Parent == null)
+					.Left.JoinAlias(x => x.Children, () => children)
+					.Left.JoinAlias(x => x.Parent, () => parent)
+					.TransformUsing(Transformers.DistinctRootEntity)
+					.Cacheable()
+					.CacheMode(CacheMode.Normal)
+					.Future();
+
+				//get values from cache
+				var roles = session
+					.QueryOver<Role>()
+					.Where(x => x.Parent == null)
+					.Left.JoinAlias(x => x.Children, () => children)
+					.Left.JoinAlias(x => x.Parent, () => parent)
+					.TransformUsing(Transformers.DistinctRootEntity)
+					.Cacheable()
+					.CacheMode(CacheMode.Normal)
+					.Future();
+
+				var result = roles.ToList();
+
+				Assert.AreEqual(3, result.Count);
+			}
+		}
+
+		[Test]
+		public void UsingHqlToFutureWithCacheAndTransformerDoesntThrow()
+		{
+			//NH-3596
+			using (var session = this.OpenSession())
+			using (session.BeginTransaction())
+			{
+				//store values in cache
+				session
+					.CreateQuery("select r from Role r left join r.Children left join r.Parent where r.Parent is null")
+					.SetResultTransformer(Transformers.DistinctRootEntity)
+					.SetCacheable(true)
+					.SetCacheMode(CacheMode.Normal)
+					.Future<Role>();
+
+				//get values from cache
```

Comment:
In other words the first future is pending and not executed before the second one gets executed. So both queries (although being the same) ends-up batched together in a single round-trip to database. So it has nothing to retrieve from cache, since only one actual sql-query is performed.

The previous test has the same issue.